lib/fluent/plugin/buffer/chunk.rb

```diff
@@ -0,0 +1,126 @@
+#
+
+require 'fluent/msgpack_factory'
+require 'fluent/plugin/buffer'
+
+require 'fileutils'
+require 'monitor'
+
+module Fluent
+  module Plugin
+    class Buffer
+      class Chunk
+        include MonitorMixin
+        include MessagePackFactory::Mixin
+
+        # Chunks has 2 part:
+        # * metadata: contains metadata which should be restored after resume (if possible)
+        #             v: {key=>value,key=>value,...} (optional)
+        #             t: tag as string (optional)
+        #             k: time slice key (optional)
+        #
+        #             id: unique_id of chunk (*)
+        #             r: number of records (*)
+        #             c: created_at as unix time (*)
+        #             m: modified_at as unix time (*)
+        #              (*): fields automatically injected by chunk itself
+        # * data: binary data, combined records represented as String, maybe compressed
+
+        # NOTE: keys of metadata are named with a single letter
+        #       to decread bytesize of metadata I/O
+
+        # TODO: CompressedPackedMessage of forward protocol?
+
+        def initialize(metadata)
+          super()
+          @unique_id = generate_unique_id
+          @metadata = metadata
+
+          @records = 0
+          @created_at = Time.now
+          @modified_at = Time.now
+        end
+
+        attr_reader :unique_id, :metadata, :created_at, :modified_at
+
+        def generate_unique_id
+          now = Time.now.utc
+          u1 = ((now.to_i * 1000 * 1000 + now.usec) << 12 | rand(0xfff))
+          [u1 >> 32, u1 & 0xffffffff, rand(0xffffffff), rand(0xffffffff)].pack('NNNN')
+        end
+
+        # data is array of formatted record string
+        def append(data)
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def commit
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def rollback
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def size
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def records
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def empty?
+          size == 0
+        end
+
+        def close
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def purge
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def read
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def open(&block)
+          raise NotImplementedError, "Implement this method in child class"
+        end
+
+        def write_to(io)
+          open do |i|
+            FileUtils.copy_stream(i, io)
+          end
+        end
+
+        def msgpack_each(&block)
+          require 'fluent/engine'
```

Comment:
Ah, it's actually nonsense requirement because it already depends on `fluent/msgpack_factory`.